counsel.el

```diff
@@ -1322,34 +1322,38 @@ files in a project.")
        (setq cmd counsel-git-grep-cmd-default)))
     (cons proj cmd)))
 
-(defun counsel--call (&rest program-and-args)
-  (let ((stderr-file (make-temp-file "counsel-call-stderr")))
+(defun counsel--call (&rest command)
+  "Synchronously call COMMAND and return its output as a string.
+COMMAND comprises the program name followed by its arguments, as
+in `make-process'.  Signal `file-error' and emit a warning if
+COMMAND fails.  Obey file handlers based on `default-directory'."
+  (let ((stderr (make-temp-file "counsel-call-stderr-")))
     (unwind-protect
          (with-temp-buffer
-           (let ((res (apply #'call-process (car program-and-args)
-                             nil (list t stderr-file)
-                             nil (cdr program-and-args))))
-             (if (= 0 res)
-                 (buffer-substring (point-min)
-                                   (- (point-max)
-                                      (if (= ?\n (char-before (point-max)))
-                                          1
-                                        0)))
-               (error "%S exited with %d\n%s"
-                      (mapconcat #'identity program-and-args " ")
-                      res
-                      (with-temp-buffer
-                        (insert-file-contents stderr-file)
-                        (buffer-string))))))
-      (delete-file stderr-file))))
+           (let ((res (apply #'process-file (car command) nil
+                             (list t stderr) nil (cdr command))))
+             (unless (eq res 0)
+               (setq res (list 'file-error
+                               (mapconcat #'identity `(,@command "failed") " ")
+                               res))
+               (lwarn 'ivy :warning "%s"
+                      (apply #'concat
+                             (error-message-string res)
+                             (when (file-readable-p stderr)
+                               (insert-file-contents stderr nil nil nil t)
+                               (list "\n" (buffer-string)))))
+               (signal (car res) (cdr res)))
+             (buffer-substring (point-min) (- (point) (if (bolp) 1 0)))))
+      (when (file-exists-p stderr)
+        (delete-file stderr)))))
 
 (defun counsel--git-grep-count-func-default ()
   "Default function to calculate `counsel--git-grep-count'."
-  (if (eq system-type 'windows-nt) 0
-    (condition-case nil
-        (let ((git-dir (counsel--call "git" "rev-parse" "--git-dir")))
-          (read (counsel--call "du" "-s" git-dir)))
-      (error 0))))
+  (or (unless (eq system-type 'windows-nt)
```

Comment:
Why Windows is excluded?  MSYS2 has [coreutils](http://github.com/Alexpux/MSYS2-packages/tree/master/coreutils) package and also available as standalone from [gnuwin32](http://gnuwin32.sourceforge.net/packages/coreutils.htm).  I think you better let it fail simply which would be consistent on either Unix or Windows if `du` is not available (what makes you think that there could not be a situation on Unix when `du` is not available?).

Reply:
> Why Windows is excluded?

Perhaps @abo-abo can explain; see 8f87f740e211a47cdcffa1b4fe086ba5240472fb.

I don't use Windows so I'm not sure what the best approach is.

Reply:
On Windows it's possible for the user to use https://git-scm.com/download/win, with no `du` available on the system. Plus process calls are a lot more expensive on Windows than on Linux. So I decided to go with piping the whole git tree into Emacs and do elisp filtering, rather than start a new process on each key press. Of course it will work poorly with a repo that has over 20k lines; it's a trade-off.

Reply:
Did you profile your statement about slow process calling on Windows?  I believe that decision is not worth the hassle that it creates, namely platform support asymmetry and, as a result, convoluted implementation and additional maintenance.  Most of those who use Emacs under Windows, will have either MSYS2 or Cygwin, which both provide `du`.  If `du` is not there, then the call will simply fail and you can still fallback to your filtering algorithm.  Otherwise, I see no reason to cut off Windows power users like this.  As you admitted, your custom filtering will not perform well on large repos anyways and 20K is nothing by the way.

Reply:
Furthermore, may I ask why do you have to do you have to call `du` on each key press?